### ISSUE TITLE
The simulations of some PowerGrids models differ based on OS and CPU

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#11935]
I am aware of the fact that different CPUs and different OSs (which means, different compiled numerical libraries) can lead to slightly different numerical results, but I thought this was something of the order of 1e-10 or less. I've come across a more dramatic case.

Consider the `PowerGrids.Examples.ENTSOE.TestCase3` model of PowerGrids 1.0.3. The model includes annotations to set `--daeMode` (hence using the IDA DAE solver, no matter what the -s flag says) and to set the tolerance and time intervals. I ran four different simulations:
1. i7-8550U CPU, Windows 10, v1.23.0-dev-265-gf870bb666a (my PC, latest nightly build with MINGW64)
2. i7-8550U CPU, Windows 10, v1.23.0-dev-265-gf870bb666a (my PC, first nightly build with UCRT64)
3. Xeon E5-2650, Ubuntu 22.04,1.23.0~dev-231-g0cd4e24 (my old Linux server, recent nightly build)
4. Ryzen 9 5950X, Ubuntu, v1.23.0-dev.266+g928b685ff4 (one of our CI test machines)

The versions of the compiler are not exactly the same, but they are all very recent nightlies and I'm not aware of any change to the code generation process. Only changes to the libraries when we moved from MINGW64 to UCRT64, which is a bit like changing the OS

These are the simulation statistics in the four cases:

Case 1.

> events
> 14 state events
> 0 time events
> solver: ida
> 829 steps taken
> 1173 calls of functionODE
> 221 evaluations of jacobian
> 32 error test failures
> 0 convergence test failures

Case 2.

> events
> 14 state events
> 0 time events
> solver: ida
> 830 steps taken
> 1168 calls of functionODE
> 219 evaluations of jacobian
> 33 error test failures
> 0 convergence test failures
> 

Case 3.

> events
> 13 state events
> 0 time events
> solver: ida
> 777 steps taken
> 1113 calls of functionODE
> 223 evaluations of jacobian
> 34 error test failures
> 0 convergence test failures

Case 4.

> events
> 13 state events
> 0 time events
> ida
> 771 steps taken
> 1118 calls of functionODE
> 232 evaluations of jacobian
> 41 error test failures
> 0 convergence test failures

The differences are significant. Even more striking is the fact that I generated reference files on my PC (Case 1.) and uploaded them to the server for verification, and it turns out that they are different enough to trigger some verification failures, e.g. [this one](https://libraries.openmodelica.org/branches/master/PowerGrids/files/PowerGrids_PowerGrids.Examples.ENTSOE.TestCase3.diff.NTHV.port.v.im.html)

 where they cause a verification failure for some signals, e.g. [NTHV.port.v.im](https://libraries.openmodelica.org/branches/master/PowerGrids/files/PowerGrids_PowerGrids.Examples.ENTSOE.TestCase3.diff.NTHV.port.v.im.html)
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/b0a0829a-728e-423a-85b0-ec9a8a823b17)

I guess this demonstrates that a tolerance of 1e-6 is really not enough for verification, not only among different tools but even with the same tool on different hw/sw architectures.

**Comment by casella** [comment]
@ceraolo, @adriguir, @marcochiaramello, @AndreaBartolini, I guess you may find this interesting 😃

**Comment by ceraolo** [comment]
> @ceraolo, @adriguir, @marcochiaramello, @AndreaBartolini, I guess you may find this interesting 😃
interesting and crazy at the same time!

**Comment by casella** [comment]
> > I would be curious to see what happens when tightening the experiment annotation  tolerance.

I will try that for sure.

@ceraolo you may also find this [ENTSO-E report](https://eepublicdownloads.entsoe.eu/clean-documents/pre2015/publications/entsoe/RG_SOC_CE/[131127_Controller_Test_Report.pdf](https://eepublicdownloads.entsoe.eu/clean-documents/pre2015/publications/entsoe/RG_SOC_CE/131127_Controller_Test_Report.pdf)) interesting, along similar lines. RTE uses it as a strong motivation in favour of using open source models and algorithms.

**Comment by adriguir** [comment]
> > > I would be curious to see what happens when tightening the experiment annotation  tolerance.
> 
> I will try that for sure.
> 
> @ceraolo you may also find this [ENTSO-E report](https://eepublicdownloads.entsoe.eu/clean-documents/pre2015/publications/entsoe/RG_SOC_CE/%5B131127_Controller_Test_Report.pdf%5D(https://eepublicdownloads.entsoe.eu/clean-documents/pre2015/publications/entsoe/RG_SOC_CE/131127_Controller_Test_Report.pdf)) interesting, along similar lines. RTE uses it as a strong motivation in favour of using open source models and algorithms.

Indeed, it has been in all my presentations for years to illustrate why we need open models and solvers :)

**Comment by ceraolo** [comment]
@casella, yourlink sem broken. Do you mean the following report?
https://eepublicdownloads.entsoe.eu/clean-documents/pre2015/publications/entsoe/RG_SOC_CE/131127_Controller_Test_Report.pdf

**Comment by ceraolo** [comment]
Entsoe report is very interesting, and is a strong indication for usage of open models and solvers as @adriguir remembers.
But open models and solvers are not enough if with them we get 3% errors as in this ticket!
That's why it is very important to find how to eliminate them.

**Comment by casella** [comment]
Remember that the tolerance is cheked on one-step-ahead error. For poorly damped systems closed to the stability limit, errors tend to accumulate, so maybe Tolerance = 1e-6 is really not tight enough. I'll try with 1e-8 as soon as I have time.

**Comment by bilderbuchi** [comment]
> But open models and solvers are not enough if with them we get 3% errors as in this ticket!
That's why it is very important to find how to eliminate them.

I can't say that I'm surprised that this happens, I encounter this problem (semi)regularly in other codebases. It could be indicative of
* loss of floating-point precision, e.g. because of 
  - insufficient data type size (`single` vs. `double`, improbable here I guess)
  - floating-point arithmetic problems with catastrophic cancellation or [accumulation of error](https://en.wikipedia.org/wiki/Kahan_summation_algorithm), etc
  - Suboptimal arranging of operations (https://herbie.uwplse.org/)
* [Numerical stability](https://nhigham.com/2020/08/04/what-is-numerical-stability/) of the used algorithms/solvers
* Plain old bugs, e.g. non-bounds-checked array access that fetches data from outside an array's memory region -- what _resides_ in that location may depend heavily on OS/compiler/version details.
* etc. etc. 😄

**Comment by casella** [comment]
@bilderbuchi I'm aware of all that, but I wouldn't expect such a significant impact. BTW, it seems mostly a Windows vs. Linux issue. Specifically
- insufficient data type size shouldn't be the case, we are (should?) be using double-precision for computations all the time
- accumulation of errors: could be, since the system is poorly damped
- numerical stability: ditto
- non-bounds-checked array access: doesn't seem likely, it would probably cause larger errors

Nice [tutorial](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html), BTW 😄

**Comment by casella** [comment]
Which includes a most disturbing statement for us Modelicans

> Incidentally, some people think that the solution to such anomalies is never to compare floating-point numbers for equality, but instead to consider them equal if they are within some error bound E. This is hardly a cure-all because it raises as many questions as it answers. What should the value of E be? If x < 0 and y > 0 are within E, should they really be considered to be equal, even though they have different signs? Furthermore, the relation defined by this rule, a ~ b  |a - b| < E, is not an equivalence relation because a ~ b and b ~ c does not imply that a ~ c.

Food for thought...

**Comment by phannebohm** [comment]
[This bachelors thesis](http://kulla.me/de/artikel/epsilon_ungefaehre_analysis/epsilon_ungefaehre_analysis.pdf) (German only) starts to develop an "epsilon-approximate" calculus. Maybe something like that is needed to have strong confidence in simulation results. Probably related to interval arithmetic, which has its own can of worms I guess...

**Comment by casella** [comment]
Vaste programme...
